### ISSUE TITLE
Compiler support and recognizion

### ISSUE THREAD

**Issue by awvwgk** [fortran-lang/fpm#223]
Came up in #220.

At some point we might have to determine which compiler is used with fpm, therefore I tried to collect some information on the various Fortran compilers around. I only have first hand access to a fraction of them, so most of these information are second hand from HPC documentations or build systems.

*Feel free to correct or add information to this table. Alternatively, we could move this table to the wiki.*

| vendor | Fortran compiler | C compiler | Module output directory | Module include directory | OpenMP | Free for OSS |
| --- | --- | --- | --- | --- | --- | --- |
| Gnu | `gfortran` | `gcc` | `-J` | `-I` | `-fopenmp` | :heavy_check_mark: |
| Intel | `ifort` | `icc` | `-module` | `-I` | `-qopenmp` | :heavy_check_mark: |
| Intel oneAPI | `ifx` | `icx` | `-module` | `-I` | `-qopenmp` | :heavy_check_mark: |
| PGI | `pgfortran` | `pgcc` | `-module` | `-I` | `-mp` | :heavy_check_mark: |
| NVIDIA | `nvfortran` | `nvc` | `-module` | `-I` | `-mp` | :heavy_check_mark: |
| LLVM flang | `flang` | `clang` | `-module` | `-I` | `-mp` | :heavy_check_mark: |
| LFortran | `lfortran` | --- | ? | ? | ? | :heavy_check_mark: |
| Lahey/Futjitsu | `lfc` | ? | `-M` | `-I` | `-openmp` | ? |
| NAG | `nagfor` | ? | `-mdir` | `-I` | `-openmp` | ? |
| Cray | `crayftn` | `craycc` | ? | ? | `-homp` | ? |
| IBM | `xlf90` | ? | ? | ? | `-qsmp` | ? |
| Oracle/Sun | ? | ? | `-moddir=` | `-M` | `-xopenmp` | ? |
| Silverfrost FTN95 | ? | ? | ? | ? | ? | ? |
| Elbrus | ? | `lcc` | `-J` | `-I` | `-fopenmp` | ? |
| Hewlett Packard | ? | ? | ? | ? | ? | discontinued |
| Watcom | ? | ? | ? | ? | ? | discontinued |
| PathScale | ? | ? | `-module` | `-I` | `-mp` | discontinued |
| G95 | ? | ? | `-fmod=` | `-I` | `-fopenmp` | discontinued |
| Open64 | ? | ? | `-module` | `-I` | `-mp` | discontinued |
| Unisys | ? | ? | ? | ? | ? | discontinued |

For now it might be sufficient to only support GCC, since we only guarantee that fpm works when compiled with GCC right now.

First class compiler support requires access to the compiler to allow testing in a CI environment, some of the commercial compilers have free versions for open source developers available.

**Comment by awvwgk** [comment]
Overview of compilers known by the meson build system can be found [here](https://mesonbuild.com/Reference-tables.html#compiler-ids), actual implementation is [here](https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/fortran.py).

**Comment by LKedward** [comment]
Awesome, thanks @awvwgk this is super useful! I've edited to include Intel differences on Windows.

**Comment by everythingfunctional** [comment]
Thanks for putting this together. It will be invaluable as we try to support more compilers.

**Comment by scivision** [comment]
* I have asked NAG to have free access for use in maintaining NAG support for Meson build system, but never got a reply. I did get replies to other email topics with NAG.
* Intel (non-oneAPI) is also free for OSS on Windows
* IBM XL is generally available [at no cost](https://www.ibm.com/products/xl-fortran-linux-compiler-power) assuming you have an IBM Power system to run it on.

**Comment by awvwgk** [comment]
I found that Ubuntu Focal (`ubuntu-20.04` image) has a [`flang` package](https://packages.ubuntu.com/source/focal/flang) available.

**Comment by scivision** [comment]
This Ubuntu flang possibly came from Debian. It might be quite out of date i.e. syntax bugs that might make issues for modern Fortran packages.

**Comment by rouson** [comment]
@awvwgk This has to be the most comprehensive listing of Fortran compilers that I've ever seen.  Bravo!

**Comment by scivision** [comment]
Yes I think this would be great for the WIki insetad of Issue. Also, PGI is now also discontinued, replaced by NVIDIA

**Comment by awvwgk** [comment]
Feel free to create a wiki page for it, the wiki should be still world-editable.

**Comment by scivision** [comment]
The wiki isn't world editable

**Comment by awvwgk** [comment]
That can be fixed. Try again.

**Comment by scivision** [comment]
Thanks, I put an updated version of this table at https://github.com/fortran-lang/fpm/wiki/Compiler-support
